examples/language_model/gpt-3/dygraph/dataset.py

```diff
@@ -0,0 +1,390 @@
+        return shuffle_idx_first
+
+    shuffle_idx_last = np.arange(
+        start=num_samples, stop=total_size, step=1, dtype=dtype_)
+    np_rng.shuffle(shuffle_idx_last)
+
+    return np.concatenate((shuffle_idx_first, shuffle_idx_last))
+
+
+def get_train_valid_test_split_(splits_string, size):
+    """ Get dataset splits from comma or '/' separated string list."""
+
+    splits = []
+    if splits_string.find(',') != -1:
+        splits = [float(s) for s in splits_string.split(',')]
+    elif splits_string.find('/') != -1:
+        splits = [float(s) for s in splits_string.split('/')]
+    else:
+        splits = [float(splits_string)]
+    while len(splits) < 3:
+        splits.append(0.)
+    splits = splits[:3]
+    splits_sum = sum(splits)
+    assert splits_sum > 0.0
+    splits = [split / splits_sum for split in splits]
+    splits_index = [0]
+    for index, split in enumerate(splits):
+        splits_index.append(splits_index[index] + int(
+            round(split * float(size))))
+    diff = splits_index[-1] - size
+    for index in range(1, len(splits_index)):
+        splits_index[index] -= diff
+    assert len(splits_index) == 4
+    assert splits_index[-1] == size
+    return splits_index
+
+
+def create_pretrained_dataset(args,
+                              input_path,
+                              data_world_rank,
+                              data_world_size,
+                              eos_id,
+                              worker_init=None,
+                              max_seq_len=1024,
+                              places=None,
+                              data_holders=None):
+    device_world_size = paddle.distributed.get_world_size()
+    device_world_rank = paddle.distributed.get_rank()
+
+    logger.info(
+        "The distributed run, total device num:{}, distinct dataflow num:{}.".
+        format(device_world_size, data_world_size))
+
+    process_datas = np.load(input_path, mmap_mode="r+", allow_pickle=True)
+    # All documment ids, extend as 1-D array.
+    sample_ids = process_datas["ids"]
+    # The len(sample_lens) num of docs
+    # The sum(sample_lens) should equal len(sample_ids)
+    sample_lens = process_datas["lens"]
+
+    splits = get_train_valid_test_split_(args.split, len(sample_lens))
+    assert len(sample_lens) >= splits[
+        -1], "The document nums should larger than max of splits, but %s < %s" % (
+            len(sample_lens), splits[-1])
+
+    def build_dataset(index, name, num_samples):
+        dataset = GPTDataset(
+            file_path=input_path,
+            build_data_file=device_world_rank == 0,
+            name="gpt_" + name,
+            max_seq_len=max_seq_len,
+            num_samples=num_samples,
+            documents=np.arange(splits[index], splits[index + 1]),
+            sample_ids=sample_ids,
+            sample_lens=sample_lens,
+            eos_id=eos_id,
+            seed=args.seed)
+
+        batch_sampler = paddle.io.DistributedBatchSampler(
```

Comment:
If we don't use shuffle, here can use
```
from paddlenlp.utils.batch_sampler import DistributedBatchSampler
```
As https://github.com/PaddlePaddle/PaddleNLP/pull/930/files

`paddlenlp.utils.batch_sampler` could save more memory.

Reply:
done， thx